firebase-database/src/androidTest/java/com/google/firebase/database/QueryTest.java

```diff
@@ -4578,22 +4578,318 @@ public void testGetWaitsForConnection() throws DatabaseException, InterruptedExc
+        appForDatabaseUrl(IntegrationTestValues.getDatabaseUrl(), UUID.randomUUID().toString());
+    FirebaseApp writeApp =
+        appForDatabaseUrl(IntegrationTestValues.getDatabaseUrl(), UUID.randomUUID().toString());
+    FirebaseDatabase readDb = FirebaseDatabase.getInstance(readApp);
+    FirebaseDatabase writeDb = FirebaseDatabase.getInstance(writeApp);
+    String refKey = UUID.randomUUID().toString();
+    DatabaseReference writeRef = writeDb.getReference().child(refKey);
+    DatabaseReference readRef = readDb.getReference().child(refKey);
+    await(
+        writeRef.setValue(new MapBuilder().put("child1", "test1").put("child2", "test2").build()));
+    ReadFuture readFuture =
+        new ReadFuture(
+            readRef,
+            events -> {
+              assertEquals(1, events.size());
+              DataSnapshot snapshot = events.get(0).getSnapshot();
+              assertEquals(2, snapshot.getChildrenCount());
+              assertEquals("test1", snapshot.child("child1").getValue());
+              assertEquals("test2", snapshot.child("child2").getValue());
+              return true;
+            });
+    readFuture.timedGet();
+    Query query = readRef.limitToFirst(1);
+    DataSnapshot snapshot = await(query.get());
+    Map<String, Object> map = new MapBuilder().put("child1", "test1").build();
+    assertEquals(map, snapshot.getValue());
+  }
+
+  @Test
+  public void testGetWithLimitToListenerOnlyFiresOnce()
+      throws DatabaseException, InterruptedException, ExecutionException, TestFailure,
+          TimeoutException {
+    FirebaseApp readApp =
+        appForDatabaseUrl(IntegrationTestValues.getDatabaseUrl(), UUID.randomUUID().toString());
+    FirebaseApp writeApp =
+        appForDatabaseUrl(IntegrationTestValues.getDatabaseUrl(), UUID.randomUUID().toString());
+    FirebaseDatabase readDb = FirebaseDatabase.getInstance(readApp);
+    FirebaseDatabase writeDb = FirebaseDatabase.getInstance(writeApp);
+    String refKey = UUID.randomUUID().toString();
+    DatabaseReference writeRef = writeDb.getReference().child(refKey);
+    DatabaseReference defaultRef = readDb.getReference().child(refKey);
+    Query readQuery = defaultRef.limitToFirst(1);
+    await(
+        writeRef.setValue(new MapBuilder().put("child1", "test1").put("child2", "test2").build()));
+    ReadFuture readFuture =
+        new ReadFuture(
+            readQuery,
+            events -> {
+              assertEquals(1, events.size());
+              DataSnapshot snapshot = events.get(0).getSnapshot();
+              assertEquals("test1", snapshot.child("child1").getValue());
+              return true;
+            });
+    readFuture.timedGet();
+    DataSnapshot snapshot = await(defaultRef.get());
+    Map<String, Object> map =
+        new MapBuilder().put("child1", "test1").put("child2", "test2").build();
+    assertEquals(map, snapshot.getValue());
+  }
+
+  @Test
+  public void testStartWithGetWithListenerOnlyFiresOnce()
```

Comment:
What about a case where we:

- Set data from writerApp
- Listen for child1
- Get child2 + child3

(disjoint listen and get results)